internal/config/config.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
```

Comment:
I am not sure if the config can be in the internal.
@mengqiy would not it impact the integration with SDK?

Reply:
I think we don't know that yet, and changing it will just be moving this folder to the `pkg` directory. I think that should not stop the PR to get merged, as the change is pretty easy to make in case we do need to export it.

Reply:
I think today is possible to use the PROJECT because it is not an internal impl. Shows reasonable for me use the Config outside of the kubebuilder project as well. E.g integration with SDK. 

So, IHMO if you really would like to move this logic for the internal and change the current design is required to check if it will not impact the current roadmap of the project and its integration with SDK first.

Reply:
We already talked about who owns the configuration file. It should not be modified by plugins: https://github.com/kubernetes-sigs/kubebuilder/pull/1198#discussion_r363992741

Moving it to an `internal` package reduces the attack surface of `kubebuilder`. In `pkg`, any third party project could import this. Moving internal machinery to `internal` packages is the only way to reduce the attack surface of `kubebuilder` and only offer to third party packages that import `kubebuilder` as a library to use the API we allowed them to.

Using `pkg/config` would be perfectly feasible too and a straight away change, lets ee what maintainers have to say about this. @mengqiy @DirectXMan12

Reply:
https://github.com/kubernetes-sigs/kubebuilder/pull/1290#discussion_r366627734 seems to suggest that they will need access to the `Config` struct so I opted for a hybrid approach. `Config` can be found under `pkg/model/config` but the `New`, `Load`, `LoadFrom` and `Save` methods are still private to guarantee that kubebuilder is the owner and the only one modifying the config file.